tq/basic_upload.go

```diff
@@ -123,7 +123,7 @@ func (a *basicUploadAdapter) DoTransfer(ctx interface{}, t *Transfer, cb Progres
 			err = errors.Wrap(err, perr.Error())
 		}
 
-		if res.StatusCode == 429 {
+		if res != nil && res.StatusCode == 429 {
```

Comment:
I think that this change is safe, since it's the only place in this block that accesses `res`, and the block returns regardless on either L129 or L132. To harden this change for the future, it might be good to make sure that we can add other checks that involve dereferencing `res`, without having to repeat the `if res != nil` bits.

What do you think about the following?

```suggestion
		if res != nil {
			// We encountered a network or similar error which caused us
			// to not receive a response at all.
			return errors.NewRetriableError(err)
		}
		
		if res.StatusCode == 429 {
```

Reply:
Yeah, that sounds like a good idea, except I think we want to check for `res == nil`. I'll push up a change with that.